stackslib/src/net/inv/nakamoto.rs

```diff
@@ -233,3 +243,559 @@ impl InvGenerator {
+}
+
+impl NakamotoTenureInv {
+    pub fn new(
+        first_block_height: u64,
+        reward_cycle_len: u64,
+        neighbor_address: NeighborAddress,
+    ) -> Self {
+        Self {
+            state: NakamotoInvState::GetNakamotoInvBegin,
+            tenures_inv: BTreeMap::new(),
+            last_updated_at: 0,
+            first_block_height,
+            reward_cycle_len,
+            neighbor_address,
+            cur_reward_cycle: 0,
+            online: true,
+            start_sync_time: 0,
+        }
+    }
+
+    /// Does this remote neighbor have the ith tenure data for the given (absolute) burn block height?
+    /// (note that block_height is the _absolute_ block height)
+    pub fn has_ith_tenure(&self, burn_block_height: u64) -> bool {
+        if burn_block_height < self.first_block_height {
+            return false;
+        }
+
+        let Some(reward_cycle) = PoxConstants::static_block_height_to_reward_cycle(
+            burn_block_height,
+            self.first_block_height,
+            self.reward_cycle_len,
+        ) else {
+            return false;
+        };
+
+        let Some(rc_tenures) = self.tenures_inv.get(&reward_cycle) else {
+            return false;
+        };
+
+        let sortition_height = burn_block_height - self.first_block_height;
+        let rc_height = u16::try_from(sortition_height % self.reward_cycle_len)
+            .expect("FATAL: reward cycle length exceeds u16::MAX");
+        rc_tenures.get(rc_height).unwrap_or(false)
+    }
+
+    /// How many reward cycles of data do we have for this peer?
+    pub fn highest_reward_cycle(&self) -> u64 {
+        self.tenures_inv
+            .last_key_value()
+            .map(|(highest_rc, _)| *highest_rc)
+            .unwrap_or(0)
+    }
+
+    /// How many blocks are represented in this inv?
+    fn num_blocks_represented(&self) -> u64 {
+        let mut total = 0;
+        for (_, inv) in self.tenures_inv.iter() {
+            total += u64::from(inv.len());
+        }
+        total
+    }
+
+    /// Add in a newly-discovered inventory.
+    /// NOTE: inventories are supposed to be aligned to the reward cycle
+    /// Returns true if we learned about at least one new tenure-start block
+    /// Returns false if not.
+    pub fn merge_tenure_inv(&mut self, tenure_inv: BitVec<2100>, reward_cycle: u64) -> bool {
+        // populate the tenures bitmap to we can fit this tenures inv
+        let learned = self
+            .tenures_inv
+            .get(&reward_cycle)
+            .map(|cur_inv| cur_inv != &tenure_inv)
+            .unwrap_or(true);
+
+        self.tenures_inv.insert(reward_cycle, tenure_inv);
+        self.last_updated_at = get_epoch_time_secs();
+        learned
+    }
+
+    /// Adjust the next reward cycle to query.
+    /// Returns the reward cycle to query.
+    pub fn next_reward_cycle(&mut self) -> u64 {
+        let query_rc = self.cur_reward_cycle;
+        self.cur_reward_cycle += 1;
```

Comment:
Just a nit: we are not particularly consistent with our +1 behaviour. Would we expect this to wrapping add? saturating add? I assume wrapping add. But I tend to avoid straight += 1

Reply:
Thanks! I'll make this a saturating add

Reply:
Just FYI you can wrap any integer type with `std::num::Saturating` ([link](https://doc.rust-lang.org/std/num/struct.Saturating.html)) to force consistent use of saturating arithmetic operations